### DISCUSSION TITLE
Future Plans For Toy

### DISCUSSION THREAD

**Discussion by Ratstail91** [krgamestudios/Toy#95] [Ideas]
First, I'd like to give a big thanks to @Add00 for their recent contributions (expanding the standard library and adding the math library).

I wanted to lay out some ideas and plans I have for Toy's future in one place, rather than having many scattered notes everywhere.

First and foremost - documentation. Now that there's a math library, it needs to be documented properly. But also, there's other elements of Toy that need to be documented, such as the game engine (which is peripherally connected to the lang, though the lang does stand on it's own). I also want to fully document the current state of the bytecode format for @hiperiondev.

Speaking of the [game engine](https://github.com/Ratstail91/Box) - I'm a gamedev at heart, so I want to use Toy for something practical. I've always wanted to use Toy to make a game which was heavily moddable, in a similar vein as something like Minecraft. So I'm also trying to work on the engine and the basics of a game too.

Other things that I wish I'd managed to finish before release, but were pushed back include:

* A threading library
* A timer library
* More random generation libraries (numbers, perlin noise, wave function collapse?)
* IO library
* Multiple return values from functions
* Interpolated strings

If there's anything you'd like to see in Toy in the future, even if it's impractical at this stage, I'd love to hear about it.

(it's 2am again, so I'll continue this train of thought tomorrow...)

**Comment by hiperiondev** [comment]
Some ideas:

- Set the values to known sizes (uintx_t, intx_t, etc) to ensure that they are respected regardless of the architecture.

- Completely separate the compiler from the interpreter: it is essential for any embedded system.

- Optimize the interpreter: This step is already more complicated but it has a very "twisted" structure which makes it difficult to understand and optimize.

**Comment by Ratstail91** [comment]
I've got this issue sitting there, actually - https://github.com/Ratstail91/Toy/issues/35

Segmenting the compiler and interpreter - I'm not strictly against it.

I'm not sure what you mean by twisted. Can you explain?

**Reply by hiperiondev** [reply]
The interpreter is not clear in its operation. It does not behave linearly like a processor. The stack is not clearly separated as such (even though it is an array). It is not easy to follow its operation nor is its execution cycle evident. A VM, in general, should be the closest thing to a finite state machine in order to optimize its speed and memory usage (with the minimum exceptions required by the language).
I don't think the interpreter follows those rules.

**Reply by hiperiondev** [reply]
On the other hand the Instruction layout must be very strict to minimize the fetch of each instruction (and due to the complexity of the interpreter it is very difficult to guess how the instructions and their arguments are composed).

**Reply by hiperiondev** [reply]
With a good explanation of the bytecode I could try to make these changes, without modifying the operation of the machine or compromising the compiler.

**Reply by hiperiondev** [reply]
After this I could start an optimization of the VM.
Surely the first thing would be in resources and then in more "esoteric" cases (depending on the infrastructure it can be much faster computed goto for efficient dispatch tables. This is already implemented by Python in your VM )

**Reply by Ratstail91** [reply]
I'm still not certain what you mean, but hopefully detailing the format will explain it well enough.

Maybe the literalCache is what's confusing you?

**Reply by hiperiondev** [reply]
Not only that, but, for example, what happens at the end readInterpreterSections (//read in the functions). It is not clear how it is referenced with the previous actions.
I suppose that when the documentation about opcodes is ready, and I can make a disassembler, everything will be clearer.

**Reply by Ratstail91** [reply]
@hiperiondev I've just added some [docs](https://toylang.com/deep-dive/developing-toy) to describe the how and why behind the literal cache.

It's not done yet, as I still need to add the opcodes, as well as a few quirks with functions, but hopefully it explains why the literal cache exists.

**Reply by hiperiondev** [reply]
Why execFnReturn reverse and back again all the stack?

**Reply by hiperiondev** [reply]
Note for OPCODE documentation

For a better understanding should be:
```
| OPCODE | ARGUMENT1 (size in bytes),  ARGUMENT2 (size in bytes) | stack before execution | stack after execution | description |
```
example:
```
| TOY_OP_OR       | (none)                                     | lit lit  | result | Logical OR       |
| TOY_OP_VAR_DECL | identifierIndex(1or 2), typeIndex (1 or 2) | lit      |    -   | Declare variable |
```

**Reply by Ratstail91** [reply]
Because it processes the stack by popping the contents off the end, and parses any identifiers found to their values. It then stores the newly parsed values in `returns`, temporarily, then returns them to the stack.

The upshot is that the order is flipped twice (leaving it unchanged), but more importantly, any identifiers that are on the stack become proper usable values.

**Reply by Ratstail91** [reply]
The opcode docs might not even fit on the page...

I'll see what I can do.

**Reply by hiperiondev** [reply]
As I think I understand the bytecode of a function is a literal and, for all intents and purposes, it is a variable. In particular it is saved in literalCache. It's right?
If so, this creates two problems:
- All the code of a function exists in RAM, and cannot reside entirely in ROM.
- Creates a complex structure when called (call/return should only touch the Program Counter and a push for the return address. In effect, a CALL is a JUMP with a return point)

This should change.

**Reply by Ratstail91** [reply]
The function is a literal, yes. However, it isn't unflattened until needed, so it is possible to rework stuff so that the `Toy_RefFunction` is pointing to a position in ROM (or even the bytecode).

> This should change.

I need to express how I'm feeling right now, and honestly, I'm feeling very pressured to complete the documentation, and to alter my code in ways that were never intended.

I've been trying to be as accommodating as possible for you, but you need to understand that I have my own things that I need to work on.

I'm still extremely eager to have the language ported to a microcontroller, and grateful that you think Toy is worth the effort. However, you also need to understand that I'm only one person. Basically, if you want some changes made to how the core program operates, then you need to get your hands dirty, and potentially fork the language to provide changes that you would like to see.

If you make changes, and open a merge request, I'd be more than happy to review it to see what benefits it would have. For the time being though, all I can provide is basic support and advice.

**Comment by Add00** [comment]
- Since I wrote the math library, I would be willing to document it. 

-  I agree I think an io library would be supper useful.

- We could use something like [FastNoiseLite](https://github.com/Auburn/FastNoiseLite/tree/master) as the basis for an efficient noise library.

- My impractical wish would be for operator overloading. Since it would be nice since if say a matrix type was introduced if users could to `matrixA + matrixB` instead of `matrixAdd(matrixA, matrixB)`

**Reply by Ratstail91** [reply]
Check the docs branch for an idea of how to write the docs up.

Once math is fully documented, feel free to start on an io library - first sketch out what you're planning to add, though, like file access, terminals, etc.

Leave the noise for now, I'll tackle that later.

**Reply by Add00** [reply]
Sure thing, also make sure to get some sleep

**Reply by Add00** [reply]
# IO Library Proposal
## Introduction

The purpose of this proposal is to outline the features and functionalities of a minimal viable IO library, which will handle file access and interaction in a straightforward and user-friendly manner. The proposed library will be implemented on top of the c [stdio](https://en.cppreference.com/w/c/io) library.

## Structures
### Mode: [string: string]

The Mode structure will be a constant dictionary containing all of the file access flags. The key-value pairs will be as follows:

- read: "r"
- write: "w"
- append: "a"
- readExtended: "r+"
- writeExtended: "w+"
- appendExtended: "a+"

The values will be in a more human-readable format, while the keys will correspond to the flags passed to the underlying C function.

### File: [string: any]

The File structure will represent a file stream and will include the following properties:

-  opaque: An opaque property representing the underlying FILE object. This property will be used for low-level operations on the file stream.
- error: A property to capture any errors that occur during file operations. This will allow users to check for errors after performing file actions.
- size: An optional property to store the size of the file. This property will be useful when dealing with binary files and for user reference.
- name: The string name of the file.

## Variables
### in: File

A file representing the standard input stream, this would correspond to `stdin`.

### out: File

A file representing the standard output stream, this would correspond to `stdout`.

### error: File

A file representing the standard error stream, this would correspond to `stderr`.

### MAX_FILENAME: int const

The largest possible size for a file name.

### MAX_FILEOPEN: int const

The largest amount of possible amount of files that can be open simultaneously.

### EOF: int const

Represents the end of a file.

## Access
### open(filename: string, mode: Mode = mode["read"]): File

The open function will allow users to open a file in the specified mode. It will take the filename and an optional mode parameter, which will default to "read" if not provided. The function will return a File object representing the opened stream. In case of failure, the error property on the returned File object will be set to indicate the specific error that occurred.

### close(file: File): bool

The close function will free the resources associated with the file stream represented by the File object. It will take the File object as input and return a boolean value indicating the success of the operation. If the file was closed successfully, it will return true; otherwise, it will return false.

## Interaction
### read(file: File, value: type): [type, bool]

The read function will attempt to read a value of the specified type from the file stream represented by the File object. Accepted value types include integer, boolean, float, and string. The function will return an array containing two elements:

- The first element will be the requested value in the requested type (e.g., if type is "integer," it will return an integer).
- The second element will be a boolean indicating whether the read operation was successful. If the read is successful, the boolean will be true; otherwise, it will be false.

### write(file: File, values...: any): bool

The write function will allow users to write one or more values to the file stream represented by the File object. Users can pass multiple values as arguments to the function, and it will write them sequentially to the file. The function will return a boolean indicating the success of the write operation. If all values were written successfully, it will return true; otherwise, it will return false.

### rename(file: File, newName: string): bool

Attempts to change the current name of the file to `newName`, returns a boolean representing the success of the operation.

### remove(file: File): bool

Will attempt to delete the file, then call `close()` on the file. Will return a boolean representing success.

## Positioning

### getPosition(file: File): int

Gets the current position in the file. Returns an int, the value will be -1 in the event of an error.

### setPosition(file: File, newPosition: int): bool

Changes the current position in the file to `newPosition`. Will return a boolean representing success.

## What is Missing?

- A way to write binary to files, I need more experience working with binary files first.
- Temporary file manipulation, I wanted to focus on permanent files first.
- Buffer manipulation, the buffers are left with the default values.
- Thread safety, I didn't really consider that when designing this.
- Directory manipulation, as that is more os dependent.
- Formatted reading and writing, this is more complicated so I want to leave it for the future.

These features are important and can always be added in a future update.

## Conclusion

By implementing the above features and functionalities, the proposed IO library will provide a simple and easy-to-use interface for file access and interaction. This minimal viable product will serve as a solid foundation for expanding the library's capabilities in the future. The library will enable developers to read and write to files and the terminal. As always feedback is greatly appreciated.

**Reply by hiperiondev** [reply]
Maybe FILE Library would be a more appropriate name. In a microcontroller IO implies many other functions (GPIO, ADC, DAC, etc)

**Reply by Add00** [reply]
Do you think fileIO would work? Since the library also has terminal interactions. Also I'm curious, how does input/output work on a microcontroller? Maybe we should make a library for that, or we can design this one to work with microcontrollers as well.

**Reply by hiperiondev** [reply]
You can see here a simple example for wifi library: https://github.com/hiperiondev/esp32_toy
It's good to define a HAL for simplify portability.
In esp32 port there is a minimal Hal for littlefs

**Reply by hiperiondev** [reply]
> Do you think fileIO would work?

it's a better name

**Reply by Ratstail91** [reply]
@Add00 This is a good start for a library. Just remember, you can always open an RFC in the discussions area, too.

This proposal does seem interesting, though I personally would pack the entire `File` into an `opaque`, and simply have accessors to check the values. For example, have a look at this:

```c
typedef struct Toy_Runner {
	Toy_Interpreter interpreter;
	const unsigned char* bytecode;
	size_t size;

	bool dirty;
} Toy_Runner;
```

This is the structure used within the runner library for running external scripts, and is stored in the root script as an `opaque`.

Another thing - you don't need a dictionary `Mode` just to specify the possible modes available - not if you document the valid values allowed.

Finally, I'm not too sure about returning a tuple form `read`; I'd prefer a function like `file.getErrors()` or something similar. So:

```toy
import fileio as fileio;

var handle: opaque = fileio.open("data:\datafile");
var value: int = handle.read(int);
var err = handle.getError();
handle.close();

if (err == null) {
    //do stuff
}
else {
    print err;
}
```

You should also look at the [drive system](https://toylang.com/c-api/drive_system_h.html) which is available in the repl folder - it's explicitly designed to help with preventing disallowed file access.

**Reply by Add00** [reply]
>  Just remember, you can always open an RFC in the discussions area, too.

Ohh okay, so something similar to [RFC: Formatted Comments As Documentation](https://github.com/Ratstail91/Toy/discussions/84)?

> I personally would pack the entire `File` into an `opaque`.

Yeah okay, this would also prevent people from setting the error state, or size to incorrect values.

> This is the structure used within the runner library for running external scripts, and is stored in the root script as an `opaque`.

Thanks, I'll definitely, take a look at that. 

> Another thing - you don't need a dictionary `Mode` just to specify the possible modes available - not if you document the valid values allowed.

I wanted to replicate an enum with the possible states, but I'll just switch `open` to use a well documented string as the second parameter.  

>  I'm not too sure about returning a tuple form read; I'd prefer a function like `file.getErrors()` or something similar. 

I could make it so that `read` returns just the type, but the type will be `null` when an error happens. Then make a `getErrors` method as you suggested.

Thanks for the feedback, it's very helpful!

**Reply by hiperiondev** [reply]
Just a comment about the libraries that depend on the Operating System or the hardware.
Before implementing the functionalities of the direct access to the resource, take the time to separate the Library - HAL, first creating the abstraction and then the port.
This makes it SO easy to switch devices without touching any of the rest of the code.
For example in https://github.com/hiperiondev/esp32_toy/blob/main/components/hal_esp32/include/hal_fs.h open, reopen, remove and rename are functions that depend on the filesystem framework and are defined as reference . In this way, changing only one compile point is directly ported to another architecture.

The structure would be
- HAL
- PORT
-- linux
-- windows
-- other_1
-- other_2
-- ...
- LIBRARY

The library ONLY refers to the HAL API. HAL does not implement any functionality (in fact it would just be headers). The functionality is implemented in the PORT.
It is common to add a hal_config.h type file for the basic configurations and, at the end, include a hal_config.h of each port.

Maybe for some common functions of an operating system it is not so obvious, but when you start to access the hardware this is critical. And when the operating system is no longer something oriented to a PC, even more so (for example FreeRTOS)

**Reply by Add00** [reply]
Thanks for the explanation! I'll have to do some more research to understand HALs, but I do want to make the library platform independent.

**Comment by hiperiondev** [comment]
> I need to express how I'm feeling right now, and honestly, I'm feeling very pressured to complete the documentation, and to alter my code in ways that were never intended.
> 

Don't worry! There's no hurry! "Should change" is not an order, it is an expression of desire!
Do not take any of my comments as pressure, but as an idea for the future.
I have to do the disassembler first so I can rethink some changes.
I await the documentation of the opcodes to be able to move forward with that, but take your time without problems.

**Reply by Ratstail91** [reply]
Thank you for understanding.

I'm starting to realize Toy is a long-term project - since I just came off of another long term project, this is actually not what I was originally planning for, but I'm rolling with it anyway.

**Reply by hiperiondev** [reply]
My comments in this discussion refer strictly to the title: Future Plans For Toy ... I mean, not now :-)
In my case, which is aimed at a device with very low resources, my interest is focused on the interpreter, which is the only thing that will reside in the microcontroller.
In this case, the internal operation is critical, but not, at least for now, neither its API nor its machine code.

**Reply by Ratstail91** [reply]
OK, thanks.

I have thought about ways of reducing the footprint of the interpreter for you - radical things like stripping out the types and such. But I'll have to leave those kinds of things up to you.

**Reply by hiperiondev** [reply]
I also understand that an embedded system like a microcontroller, with very low resources and close contact between hardware and firmware, is an alien world for most programmers. I'm used to dealing with this.
The main problem is that almost all programmers base their developments on the idea that memory can be used as if there were no tomorrow. Things like garbage collector, free dynamic structures, automatic heap growth are very dangerous in this context.
If it is taken into account that a microcontroller can drive machinery where people can be injured, certain practices are not directly acceptable.
Take any comment of mine from this point of view, and nothing more.

**Reply by hiperiondev** [reply]
And for more problems: my English is quite bad. Maybe I don't express myself correctly sometimes.
Please feel free to correct me whenever you see something like this. It doesn't offend me, on the contrary, it helps me learn!

**Reply by Ratstail91** [reply]
Your English is pretty good for the most part, nothing to worry about there. And you're right - I've never worked on a microcontroller, so I'll delegate to you for advice on that.

**Comment by Ratstail91** [comment]
@Add00 Let me know roughly when you think the IO library will be ready - I'm gearing up for 1.3, but I'd like your lib to be included in that. No rush, of course.

(Bumping to 1.3 is because I got the short-circuiting to work, finally.)

**Reply by Add00** [reply]
Hmm, I think I'll be mostly done by Wednesday this week. I could use some help with setting up the [makefile](https://github.com/Ratstail91/Toy/commit/2fa6b7ee23247b39db23756d99afafba60fca435) for the hal if you or @hiperiondev. Could help me build it as a library and include it as part of the toy build that would be really helpful. Since I don't have a lot of experaince with make I have been having trouble configuring it.

**Comment by Ratstail91** [comment]
I'm gonna be honest - until I googled "HAL" just now I wasn't entirely following what you and @hiperiondev were talking about, because I'm a big dumb dummy lol.

OK, first, let met have a proper look at your efforts so far... Oh wow. How did you come up with this design for the HAL @Add00? Making a structure which holds function pointers is very creative, though it's also very OO-centric.

OK, so... I'm not sure what your intentions are, tbh. If you've got the Toy lib, it needs to wrap the lower-level functions somehow...

```c
static int nativeOpen(Toy_Interpreter* interpreter, Toy_LiteralArray* arguments) {
  //TODO: how do I get from here to the file system?
}
```

You'll have to forgive me, as I'm not the most experienced working directly with other coders.

---

My original thoughts were that @hiperiondev would code a custom library that I would never (need to) see, while conforming to a kind of agreed-upon API that was drop-in compatible with the library that @Add00 writes for the canonical branch of Toy. But this doesn't seem to be the way we're going. What are your plans for this?

Also, does the drive system fit into this anywhere? The drive system wasn't an arbitrary feature, it's actually there to allow for players to safely drop mods into to the Box game engine without needing to worry about potential hard drive access. They should be able to just copy/paste files into a directory, eventually.

In addition, it's also there to add my own kind of HAL, since I'm actually [a published switch developer](https://www.nintendo.com/store/products/candy-raid-the-factory-switch/), and I'd like to port the Game + Box + Toy to the switch eventually, even if just for fun. My plan was to use some \#ifdefs to swap out code within the libraries when I got around to it.

So yeah, basically, what are your plans for the IO library?

**Reply by Add00** [reply]
Sorry for the delayed response. My computer took a short drop down a long flight of stairs. Putting it out of commission for a few days. 

> until I googled "HAL" just now I wasn't entirely following what you and hiperiondev were talking about

It's okay, I wasn't quite sure what they were till hiperiondev mentioned them. XD

> OK, first, let met have a proper look at your efforts so far... Oh wow. How did you come up with this design for the HAL Add00? Making a structure which holds function pointers is very creative, though it's also very OO-centric.

Honestly, before this I never made an HAL before, so I did some research and the OOP approach seemed the easiest to me. Though I'm not too sure if this it is the best way to make one.

> You'll have to forgive me, as I'm not the most experienced working directly with other coders.

No need to apologize! I'm a very inexperienced programmer who doesn't know a lot about with open source contributing.  It looks like I should have been more clear in communicating

> My original thoughts were that hiperiondev would code a custom library that I would never (need to) see, while conforming to a kind of agreed-upon API that was drop-in compatible with the library that Add00 writes for the canonical branch of Toy. But this doesn't seem to be the way we're going. What are your plans for this?

Ohhhh I didn't realize that this was the approach you were aiming for. I thought that there would be one io library which would handle all platforms internally. I think this approach of having a canonical one plus an agreed interface would be much easier to handle.

> Also, does the drive system fit into this anywhere? The drive system wasn't an arbitrary feature, it's actually there to allow for players to safely drop mods into to the Box game engine without needing to worry about potential hard drive access. They should be able to just copy/paste files into a directory, eventually.

I do plan on incorporating the drive system, I didn't have a chance since I was researching and making the HAL.

> In addition, it's also there to add my own kind of HAL, since I'm actually [a published switch developer](https://www.nintendo.com/store/products/candy-raid-the-factory-switch/), and I'd like to port the Game + Box + Toy to the switch eventually, even if just for fun. My plan was to use some #ifdefs to swap out code within the libraries when I got around to it.

Woah that really cool! I think that should still be possible.

> So yeah, basically, what are your plans for the IO library?

Now that I have a better idea of what was actually expected. I think I will drop the HAL completely and just focus on building the fileio library using the drive system to handle paths, with the feedback in this [post](https://github.com/Ratstail91/Toy/discussions/95#discussioncomment-6634275). 

I didn't mean to cause so much confusion, but now that I've read your comment I have a much better idea of what was actually intended. I'll definitely communicate more going forward to prevent more confusion.

**Reply by Ratstail91** [reply]
No worries about the delay. Laptops can be problematic at the best of times lol.

My theory is that different users would need different libraries for different situations - so I tend to plan it so that there's no guarantee that any one library will be available (though standard is a bit special in that regard).

I'm used to leading small teams on occasion (though never via github like this before), so I should've known to be more open with my ideas - my bad. At the same time though, please feel free to critique my ideas - don't take them as gospel, if you feel like there's something you'd like to add.

Don't worry about causing confusion - this always happens in new teams, and is an expected part of the plan.

Finally, I'd like to add that no one has any obligations to this project - I'll assign you a task if you ask for one, but this is ultimately a project written for fun, so lets have some fun with it!

**Reply by Add00** [reply]
I completely agree! To have fun and learn are why I wanted to contribute to this project in the first place.

**Comment by hiperiondev** [comment]
Hey!!! I did not disappear! I just have a lot of work.

Regarding the HAL, I understand that the design is complicated, especially if you never had contact with it.

You can see some well organized examples here:

https://github.com/FreeRTOSHAL/fhal-driver

https://www.chibios.org/dokuwiki/doku.php?id=chibios:products:hal:start

http://www.wsn.agh.edu.pl/download/public/halfred/0.2.0/

**Reply by Ratstail91** [reply]
> I just have a lot of work.

No worries!

Regarding the HAL - as I understand it, the internals of the HAL change based on platform. What I'm trying to convey above is that the library itself might be better off handling the HAL's role.

For example - if there's a core library called `lib_fileio`, you could have a specialized library called `lib_fileio_esp32` which replaces the former by simply changing which hook gets injected into the interpreter on startup. It would be a drop-in replacement, you see?

**Reply by hiperiondev** [reply]
It could be an option. In fact it is more or less what a HAL does. But for it to make sense, the library references should be immutable, that is, refer to an API (the HAL header). The port should only own the API implementation.
To summarize: no matter what the hardware or operating system, the Toy library should never be modified or injected with external code, and the header it references should never change (for example: suppose that for two different architectures there are two different ways of doing fopen, in that case the library should have only one access to the HAL API - fopen_hal()? - exist fopen_hal in the API entry - toy_hal.h - and its implementation in each port, which would be the responsibility of whoever wants implement each infrastructure).
The idea of having a HAL is that the specific code of each access to the hardware/operating system should NEVER be injected into the application and that it be transparent to modify it.

In fact you are already using this without knowing it :-)
Every time you make use of any access to a driver of the operating system, it implements an API to access the hardware that is, effectively, the HAL of the system.

**Reply by hiperiondev** [reply]
For example, when you use a graphics library (opengl, sdl, etc) you use an API that generates "something" (a primitive, a motion function, etc). This library accesses the graphic hardware to do it and this hardware can vary a lot, from an accelerator board to a simple display that can only turn points on and off and nothing else. You never have access to the hardware or care about it, in fact you are completely unaware of how it works and the library can work on many different hardware that work in custom ways. That's why the library has an immutable HAL implementation and its implementation depends on each hardware. Every time you want to port that library to a new video card, ONLY the implementation (port) is modified and nothing else from the library is touched. HAL API and Library API are, and should be, immutable

**Reply by Ratstail91** [reply]
I'm busy today and tomorrow, but we'll see how it works out once Add00 finishes his new io library (sorry, not too focused today).

**Comment by Ratstail91** [comment]
I'll be joining a game jam, which starts at 2am tonight - I'll be using Toy and Box...

**Reply by Add00** [reply]
This sounds, cool, which game jam was it for?

**Reply by Ratstail91** [reply]
the game wisdom jam, but IDK if I'll finish, as I don't want to push myself too hard - https://itch.io/jam/game-wisdom-game-jam

**Reply by Add00** [reply]
The jam looks interesting, even if you don't finish you can still learn and enjoy the experience!

**Comment by Ratstail91** [comment]
Sorry, been kind of slacking due to irl stuff. Will try coding tonight, maybe.
